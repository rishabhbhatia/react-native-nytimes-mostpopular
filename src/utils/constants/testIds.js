/* -------------------------------------------------------------------------- */
/*                  testIDS of application are declared here                  */
/* -------------------------------------------------------------------------- */

/* ------------------------------ articles screen ------------------------------ */
export const TEST_ID_ARTICLES_SCREEN_LIST = "articlesScreenList";
export const TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW = "articleDetailsScreenWebview";
