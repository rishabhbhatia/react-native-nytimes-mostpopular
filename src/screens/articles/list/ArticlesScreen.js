import React from "react";
import {
  SafeAreaView, View, FlatList, Alert, ActivityIndicator,
} from "react-native";

import Icon from "react-native-vector-icons/FontAwesome";

import ArticleItem from "./item/ArticleItem";

import styles from "./styles";
import { TEST_ID_ARTICLES_SCREEN_LIST } from "../../../utils/constants/testIds";

/* -------------------------------------------------------------------------- */
/*                 Screen to display list of popular articles                 */
/* -------------------------------------------------------------------------- */
export default class ArticlesScreen extends React.Component {
  /* ------------------------ custom header properties ------------------------ */
  static navigationOptions = {
    title: "NY Times Most Popular",
    headerStyle: {
      backgroundColor: "#47e4c2",
    },
    headerTintColor: "#fff",
    headerTitleStyle: {
      fontWeight: "bold",
    },
    headerLeft: (
      <Icon style={styles.icon} name="bars" size={20} color="#fff" />
    ),
    headerRight: (
      <View style={styles.headerRightContainer}>
        <Icon style={styles.icon} name="search" size={20} color="#fff" />
        <Icon style={styles.icon} name="ellipsis-v" size={20} color="#fff" />
      </View>
    ),
    headerBackTitle: null,
    headerTruncatedBackTitle: null,
  };

  state = {
    articles: [],
    totalArticlesCount: 0,
    showFooter: false,
  };

  /* ------------- once component is loaded fetch list of articles ------------ */
  componentDidMount() {
    this.fetchArticles();
  }

  /* ----------- load most popular articles and push them into state ---------- */
  fetchArticles = () => fetch("https://api.nytimes.com/svc/mostpopular/v2/mostviewed/all-sections/7.json?api-key=K3HkXGaTBgZfXIb1COHeJkM5RvWjurlo")
    .then(response => response.json())
    .then((responseJson) => {
      const { results: articles, num_results: totalArticlesCount } = responseJson;
      this.setState({ articles, totalArticlesCount });
      return responseJson;
    })
    .catch(() => {
      Alert.alert("Something went wrong, try again later");
    })

  /* ------------------ when user clicks on any article cell ------------------ */
  onArticleSelected = (article) => {
    const { navigation } = this.props;
    const { url: articleUrl } = article;
    navigation.navigate("ArticleDetails", {
      articleUrl,
    });
  }

  /* -------------------------------------------------------------------------- */
  /*                   pagination to load next set of articles                  */
  /* -------------------------------------------------------------------------- */
  // eslint-disable-next-line consistent-return
  loadMoreArticles = () => {
    const { articles, totalArticlesCount } = this.state;

    /* ----------------- if more articles exist in our database; ---------------- */
    /* -------------- load them and add to current set of articles -------------- */
    if (articles.length < totalArticlesCount) {
      this.setState({
        showFooter: true,
      });

      /* istanbul ignore next */
      return fetch(`https://api.nytimes.com/svc/mostpopular/v2/mostviewed/all-sections/7.json?api-key=K3HkXGaTBgZfXIb1COHeJkM5RvWjurlo&offset=${articles.length}`)
        .then(response => response.json())
        .then((responseJson) => {
          const { results: nextArticles, num_results: nextTotalArticlesCount } = responseJson;
          this.setState({
            articles: [...articles, ...nextArticles],
            totalArticlesCount: nextTotalArticlesCount,
            showFooter: false,
          });
          return responseJson;
        })
        .catch(() => {
          this.setState({
            showFooter: false,
          }, () => {
            Alert.alert("Something went wrong, try again later");
          });
        });
    }
  }

  /* ----------- article id defines each unique element in the list ----------- */
  keyExtractor = item => `${item.id}`;

  renderFooter = () => {
    const { showFooter } = this.state;
    if (showFooter) {
      return (
        <View style={styles.footer}>
          <ActivityIndicator size="small" />
        </View>
      );
    }
    return null;
  }


  renderEmptyComponent = (
    <View style={styles.emptyComponentContainer}>
      <ActivityIndicator size="large" />
    </View>
  );

  /* ------------------ adding a seperator between each cell ------------------ */
  renderArticlesItemSeperator = () => <View style={styles.articleItemSeperatorContainer} />

  /* ------------------------- renders an article cell ------------------------ */
  renderArticleItem = ({ item: article }) => {
    const props = {
      article,
      onPress: () => this.onArticleSelected(article),
    };
    return ArticleItem(props);
  };

  render() {
    const { articles } = this.state;
    return (
      <SafeAreaView style={styles.container}>
        <FlatList
          testID={TEST_ID_ARTICLES_SCREEN_LIST}
          keyExtractor={this.keyExtractor}
          data={articles}
          showsVerticalScrollIndicator={false}
          onEndReached={this.loadMoreArticles}
          contentContainerStyle={styles.articlesListContainer}
          renderItem={this.renderArticleItem}
          ItemSeparatorComponent={this.renderArticlesItemSeperator}
          ListFooterComponent={this.renderFooter}
          ListEmptyComponent={this.renderEmptyComponent}
        />
      </SafeAreaView>
    );
  }
}
