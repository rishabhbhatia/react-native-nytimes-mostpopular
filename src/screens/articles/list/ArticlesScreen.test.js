import React from "react";
import { shallow } from "enzyme";
import fetchMock from 'fetch-mock';

import ArticlesScreen from "./ArticlesScreen";
import { TEST_ID_ARTICLES_SCREEN_LIST } from "../../../utils/constants/testIds";

const navigation = { navigate: () => { } };

const articlesJson = require('../../../../assets/articles.json');
const articles = articlesJson.results;

/* -------------------------------------------------------------------------- */
/*                                  UI Tests                                  */
/* -------------------------------------------------------------------------- */
describe("ArticlesScreen", () => {
  /* -------------------------------- snapshot -------------------------------- */
  it("matches snapshot", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen />);
    expect(articlesScreen).toMatchSnapshot();
  });

  /* ----------------------------- articles list ----------------------------- */
  it("articles list is visible", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    expect(articlesList.exists()).toBeTruthy();
  });

  it("article list key extractor sets unique key", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    expect(articlesList.props().keyExtractor(articles[0])).toBeTruthy();
  });

  it("article list onEndReached exists", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    articlesList.props().onEndReached();
    expect(articlesList.props().onEndReached).toBeInstanceOf(Function);
  });

  it("article list onEndReached works with empty articles and total articles count in state", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);

    articlesScreen.setState({
      articles: [],
      totalArticlesCount: 20
    }, () => {
      articlesScreen.update()
      const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
      articlesList.props().onEndReached();
      expect(articlesList.props().onEndReached).toBeInstanceOf(Function);
    });
  });

  it("article list render item renders with empty article value as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    expect(articlesList.props().renderItem({})).toMatchSnapshot();
  });

  it("article list render item renders with valid article value as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    expect(articlesList.props().renderItem({ item: articles[0] })).toMatchSnapshot();
  });

  it("article list render item on press exists", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    articlesList.props().renderItem({ item: articles[0] }).props.onPress();
    expect(articlesList.props().renderItem({ item: articles[0] }).props.onPress).toBeInstanceOf(Function);
  });

  it("article list item seperator component renders as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
    expect(articlesList.props().ItemSeparatorComponent()).toMatchSnapshot();
  });

  it("article list footer component renders with showFooter true as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);

    articlesScreen.setState({
      showFooter: true
    }, () => {
      articlesScreen.update();
      const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
      expect(articlesList.props().ListFooterComponent()).toMatchSnapshot();
    });
  });

  it("article list footer component renders with showFooter false as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);

    articlesScreen.setState({
      showFooter: false
    }, () => {
      articlesScreen.update();
      const articlesList = articlesScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLES_SCREEN_LIST);
      expect(articlesList.props().ListFooterComponent()).toMatchSnapshot();
    });
  });

  it("article screen state updates as expected", () => {
    expect.assertions(1);
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    articlesScreen.setState({
      articles
    }, () => {
      articlesScreen.update();
      expect(articlesScreen.state('articles')).toEqual(articles);
    });
  });

  it("fetch article api loads the data", (done) => {
    const articlesScreen = shallow(<ArticlesScreen navigation={navigation} />);
    fetchMock.get('*', articlesJson)
    articlesScreen.instance().fetchArticles().then(res => {
      expect(res).toEqual(articlesJson);
      done();
    })
    fetchMock.reset();
  });
});
