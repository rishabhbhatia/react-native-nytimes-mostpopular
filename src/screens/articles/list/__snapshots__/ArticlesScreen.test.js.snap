// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlesScreen article list footer component renders with showFooter false as expected 1`] = `null`;

exports[`ArticlesScreen article list footer component renders with showFooter true as expected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 50,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`ArticlesScreen article list item seperator component renders as expected 1`] = `
<View
  style={
    Object {
      "marginTop": 15,
    }
  }
/>
`;

exports[`ArticlesScreen article list render item renders with empty article value as expected 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.2,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#969696",
            "borderRadius": 20,
            "height": 40,
            "width": 40,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 0.6,
          "justifyContent": "center",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={2}
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "500",
            "textAlign": "left",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#9e9e9e",
            "fontSize": 12,
            "fontWeight": "400",
            "marginTop": 10,
            "textAlign": "left",
          }
        }
      >
        
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginTop": 2,
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="#969696"
          name="calendar"
          size={10}
        />
        <Text
          style={
            Object {
              "color": "#9e9e9e",
              "fontSize": 10,
              "fontWeight": "400",
              "marginLeft": 5,
              "textAlign": "right",
            }
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.2,
          "justifyContent": "center",
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="#c9c7c7"
        name="chevron-right"
        size={15}
      />
    </View>
  </View>
</TouchableWithoutFeedback>
`;

exports[`ArticlesScreen article list render item renders with valid article value as expected 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.2,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#969696",
            "borderRadius": 20,
            "height": 40,
            "width": 40,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 0.6,
          "justifyContent": "center",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={2}
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "500",
            "textAlign": "left",
          }
        }
      >
        Teenager Accused of Rape Deserves Leniency Because He’s From a ‘Good Family,’ Judge Says
      </Text>
      <Text
        style={
          Object {
            "color": "#9e9e9e",
            "fontSize": 12,
            "fontWeight": "400",
            "marginTop": 10,
            "textAlign": "left",
          }
        }
      >
        By LUIS FERRÉ-SADURNÍ
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginTop": 2,
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="#969696"
          name="calendar"
          size={10}
        />
        <Text
          style={
            Object {
              "color": "#9e9e9e",
              "fontSize": 10,
              "fontWeight": "400",
              "marginLeft": 5,
              "textAlign": "right",
            }
          }
        >
          2019-07-02
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.2,
          "justifyContent": "center",
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="#c9c7c7"
        name="chevron-right"
        size={15}
      />
    </View>
  </View>
</TouchableWithoutFeedback>
`;

exports[`ArticlesScreen matches snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "rgb(250, 250, 250)",
      "flex": 1,
    }
  }
>
  <FlatList
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="large"
        />
      </View>
    }
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "paddingVertical": 15,
      }
    }
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    testID="articlesScreenList"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</SafeAreaView>
`;
